lib/driver-cli.js

```diff
@@ -0,0 +1,249 @@
+      // drivers installed via npm can include versions or tags after the '@'
+      // sign, so check for that
+      const [name, pkgVer] = installSpec.split('@');
+      let pkgName;
+
+      if (installType === INSTALL_TYPE_NPM) {
+        // if we're installing a named package from npm, we don't need to check
+        // against the appium driver list; just use the installSpec as is
+        pkgName = name;
+      } else {
+        // if we're installing a named appium driver (like 'xcuitest') we need to
+        // dereference the actual npm package ('appiupm-xcuitest-driver'), so
+        // check it exists and get the correct package
+        const knownNames = Object.keys(KNOWN_DRIVERS);
+        if (!_.includes(knownNames, name)) {
+          const msg = `Could not resolve driver; are you sure it's in the list ` +
+                      `of supported drivers? ${JSON.stringify(knownNames)}`;
+          throw new Error(msg);
+        }
+        pkgName = KNOWN_DRIVERS[name];
+        // given that we'll use the install type in the driver json, store it as
+        // 'npm' now
+        installType = INSTALL_TYPE_NPM;
+      }
+
+      driverData = await this.installViaNpm({driver, pkgName, pkgVer});
+    }
+
+    const {driverName} = driverData;
+    delete driverData.driverName;
+
+    if (this.config.isInstalled(driverName)) {
+      throw new Error(`A driver named '${driverName}' is already installed. ` +
+                      `Did you mean to update? 'appium driver update'. See ` +
+                      `installed drivers with 'appium driver list --installed'.`);
+    }
+
+    driverData.installType = installType;
+    driverData.installSpec = installSpec;
+    await this.config.addDriver(driverName, driverData);
+
+    // log info for the user
+    log(this.json, `Driver ${driverName}@${driverData.version} successfully installed`.green);
+    log(this.json, `- automationName: ${driverData.automationName.green}`);
+    log(this.json, `- platformNames: ${JSON.stringify(driverData.platformNames).green}`);
+
+    return this.config.installedDrivers;
+  }
+
+  async installViaNpm ({driver, pkgName, pkgVer}) {
+    let installPath = pkgName;
+    if (!pkgName) {
+      pkgName = driver;
+      installPath = path.posix.basename(pkgName);
+    }
+    const npmSpec = `${pkgName}${pkgVer ? '@' + pkgVer : ''}`;
+    const specMsg = npmSpec === driver ? '' : ` using NPM install spec '${npmSpec}'`;
+    const msg = `Installing '${driver}'${specMsg}`;
+    try {
+      const pkgJsonData = await spinWith(this.json, msg, async () => (
+        await this.npm.installPackage(path.resolve(this.config.appiumHome, installPath), pkgName, pkgVer)
+      ));
+      const driverData = this.getDriverFields(pkgJsonData);
+      driverData.installPath = installPath;
+      return driverData;
+    } catch (err) {
+      throw new Error(`Encountered an error when installing package: ${err.message}`);
+    }
+  }
+
+  getDriverFields (pkgJsonData) {
```

Comment:
like I said, some documentation on what the method accepts and returns would be super nice here